### ISSUE TITLE
feature request: Option to control GUI update speed

### ISSUE THREAD

**Issue by martin-lueders** [VCVRack/Rack#183]
OS (if applicable): MAC
Version (or "dev" if compiling from source): 0.3.2

I noticed that some of my patches struggle on my Mac (Macbook Pro 2011 model) if the GUI is open. Once I close the GUI, the glitches disappear. My patch uses a lot of modules with flashing LEDs, so I guess that the GUI update poses a high strain on the CPU.

Would it be possible to control the GUI update speed through a switch in the status line?
That might help to get a bit more mileage on older computers.

Cheers,
Martin

**Comment by AndrewBelt** [comment]
The refresh rate should be vsynced to your monitor's rate, which is around 60FPS. (If it's going higher, that's a bug.) If your GPU is not capable of tesselating the vector graphics, the software renderer will kick in, which consumes lots of CPU. There are a few optimizations better than simply halving the refresh rate, but I'll get around to that probably after v1.0 releases.

Could you describe "glitches"? If by "close the GUI" you mean minimize, I have turned off rendering when minimized, so this reinforces the fact that a lot of time is spent in the main (graphics) thread.

**Comment by martin-lueders** [comment]
With glitches I mean simply dropouts in the sound. The typical sign if the system is not fast enough to fill the buffers. When I changed the buffersize to 1024 it improved, but still had some dropouts when the GUI was open.

How can I figure out whether my Mac is able to render in hardware, or whether software is kicking in? I assume it's software, as my PC can easily handle the same patch.

**Comment by Strum** [comment]
As a matter of interest, is it possible to take processing of gui controls out of the step function and deal with them at a lower rate to reduce processing load?

It seems unnecessary to be processing controls every sample step when a much lower refresh rate would be fine.

**Comment by AndrewBelt** [comment]
It's important to call `step()` before `draw()`. Both functions are essentially "free". The bottleneck is the `nvgEndFrame()` call in gui.cpp, which takes drawlist build in `draw()` and tesselates it in a shader.

![2017-10-06-235349_2560x1422_scrot](https://user-images.githubusercontent.com/338179/31304581-0db67d88-aaf2-11e7-9d8f-b010e59abd34.png)

**Comment by Strum** [comment]
Thanks Andrew for your prompt reply.

I'm not sure I really understand the graphic but I think it means that the actual audio processing is a tiny part of the processor time in comparison to the the gui so it doesn't matter right?

had interest in the irc channel already btw, got 3 others who seem keen to hang.

**Comment by AndrewBelt** [comment]
It's a [Flame Graph](http://www.brendangregg.com/flamegraphs.html).
That's actually just the main (graphics) thread. Here's the audio thread.
![2017-10-06-235349_2560x1422_scrot](https://user-images.githubusercontent.com/338179/31304812-ee1c3aba-aaf8-11e7-8f94-26092c407a67.png)
As you can see, most of the time is spent in libsamplerate's `src_process()`. I've opened an issue to fix this. #194

I'll check out the IRC channel when I have time.

**Comment by Strum** [comment]
Thanks, I really appreciate your help.

Look forward to seeing you on irc..

**Comment by Strum** [comment]
Flame graphs, git, so much to learn, so little time.

**Comment by Strum** [comment]
These flame graphs are a great way to get an idea of how a program works and find where to start to understand the overall architecture of a program.

**Comment by AndrewBelt** [comment]
Yeah, if you don't profile your code, any optimization you do is a pre-optimization and likely a waste of time. It allows me to completely ignore performance until I actually need it.

**Comment by Strum** [comment]
Yes but they are also useful for more than just profiling to someone trying to understand how a complex program works.
